src/org/labkey/remoteapi/assay/Protocol.java

```diff
@@ -0,0 +1,415 @@
+
+    private List<String> _availableDetectionMethods;
+    private String _selectedDetectionMethod;
+
+    private Map<String, String> _availableMetadataInputFormats;
+    private String _selectedMetadataInputFormat;
+
+    private List<String> _availablePlateTemplates;
+    private String _selectedPlateTemplate;
+
+    private Map<String, String> _protocolParameters;
+    private List<String> _protocolTransformScripts;
+
+    public Protocol()
+    {
+        super(null);
+    }
+
+    public Protocol(JSONObject json)
+    {
+        super(json);
+
+        _protocolId = (Long)json.get("protocolId");
+        _name = (String)json.get("name");
+        _description = (String)json.get("description");
+        _providerName = (String)json.get("providerName");
+
+        if (json.get("domains") instanceof JSONArray)
+        {
+            for (Object domain : ((JSONArray)json.get("domains")))
+                _domains.add(new Domain((JSONObject)domain));
+        }
+
+        if (json.containsKey("allowBackgroundUpload"))
+            _allowBackgroundUpload = (Boolean)json.get("allowBackgroundUpload");
+        if (json.containsKey("backgroundUpload"))
+            _backgroundUpload = (Boolean)json.get("backgroundUpload");
+        if (json.containsKey("allowEditableResults"))
+            _allowEditableResults = (Boolean)json.get("allowEditableResults");
+        if (json.containsKey("editableResults"))
+            _editableResults = (Boolean)json.get("editableResults");
+        if (json.containsKey("editableRuns"))
+            _editableRuns = (Boolean)json.get("editableRuns");
+        if (json.containsKey("saveScriptFiles"))
+            _saveScriptFiles = (Boolean)json.get("saveScriptFiles");
+        if (json.containsKey("qcEnabled"))
+            _qcEnabled = (Boolean)json.get("qcEnabled");
+        if (json.containsKey("allowQCStates"))
+            _allowQCStates = (Boolean)json.get("allowQCStates");
+        if (json.containsKey("allowSpacesInPath"))
+            _allowSpacesInPath = (Boolean)json.get("allowSpacesInPath");
+        if (json.containsKey("allowTransformationScript"))
+            _allowTransformationScript = (Boolean)json.get("allowTransformationScript");
+        if (json.containsKey("autoCopyTargetContainerId"))
+            _autoCopyTargetContainerId = (String)json.get("autoCopyTargetContainerId");
+
+        if (json.get("availableDetectionMethods") instanceof JSONArray)
+        {
+            for (Object detectionMethod : (JSONArray)json.get("availableDetectionMethods"))
+                _availableDetectionMethods.add((String)detectionMethod);
+        }
+        if (json.containsKey("selectedDetectionMethod"))
+            _selectedDetectionMethod = (String)json.get("selectedDetectionMethod");
+
+        if (json.containsKey("availableMetadataInputFormats"))
+        {
+            _availableMetadataInputFormats = (HashMap<String,String>)json.get("availableMetadataInputFormats");
```

Comment:
Does this casting to a HashMap directly here work? I would have expected you to need to cast it to a JSONObject and then iterate through it to build up the map yourself.

Reply:
JSONObject subclasses HashMap, so this works. However, it might be a little clearer to call getJSONObject() instead of get(). Of course, JSONObject is a Map<String, Object>, not Map<String, String>, so it might need a little coercion since this code is blindly assuming that it will only contain Strings for values.